File: P/1769_Minimum_Number_of_Operations_to_Move_All_Balls_to_Each_Box.java
class Solution {
    public int[] minOperations(String boxes) {
        int[] answerLtoR = new int[boxes.length()];
        int count = 0;
        for(int i = 0; i < boxes.length() - 1; ++i) {
            if(boxes.charAt(i) == '1')
                count++;
            answerLtoR[i + 1] += answerLtoR[i] + count;
        }
        
        int[] answerRtoL = new int[boxes.length()];
        count = 0;
        for(int i = boxes.length() - 1; i > 0; --i) {
            if(boxes.charAt(i) == '1')
                count++;
            answerRtoL[i - 1] += answerRtoL[i] + count;
        }
        
        int[] answer = new int[boxes.length()];
        for(int i = 0; i < boxes.length(); ++i) {
            answer[i] = answerLtoR[i] + answerRtoL[i];
        }
        
        return answer;
    }
}

//Brute Force
class Solution {
    public int[] minOperations(String boxes) {
        int[] answer = new int[boxes.length()];
        char[] box = boxes.toCharArray();
        
        for(int cur = 0; cur < box.length; ++cur) {
            for(int other = 0; other < box.length; ++other) {
                if(box[other] == '1' && cur != other ) {
                    answer[cur] += Math.abs(cur - other); 
                }
            }
        }
        return answer;
    }
}

//Brute Force
class Solution {
    public int[] minOperations(String boxes) {
        int[] answer = new int[boxes.length()];
        
        for(int i = 0; i < boxes.length(); ++i) {
            if(boxes.charAt(i) == '1') {
                for(int j = 0; j < boxes.length(); ++j) {
                    answer[j] += Math.abs(i - j);
                }
            }
        }
        return answer;
    }
}

//Brute Force
class Solution {
    public int[] minOperations(String boxes) {
        int[] answer = new int[boxes.length()];
        
        List<Integer> ones = new ArrayList<>();
        for(int i = 0; i < boxes.length(); ++i) {
            if(boxes.charAt(i) == '1') {
                ones.add(i);
            }
        }
        
        for(int i = 0; i < boxes.length(); ++i) {
            for(int j = 0; j < ones.size(); ++j) {
                answer[i] += Math.abs(ones.get(j) - i);
            }
        }
        return answer;
    }
}

File: 주영/Any/1769. Minimum Number of Operations to Move All Balls to Each Box.js
var minOperations = function (boxes) {
  let answer = [];
  for (let i = 0; i < boxes.length; i++) {
    let sum = 0;
    for (let j = 0; j < boxes.length; j++) {
      if (boxes[j] === '1') {
        sum += Math.abs(i - j);
      }
    }
    answer.push(sum);
  }
  return answer;
};

var minOperations = function (boxes) {
  const len = boxes.length;
  const leftToRightDp = new Array(len);
  const rightToLeftDp = new Array(len);

  (leftToRightDp[0] = 0), (rightToLeftDp[len - 1] = 0);

  let oneCount = boxes[0] === '1' ? 1 : 0;

  for (let i = 1; i < len; i++) {
    leftToRightDp[i] = leftToRightDp[i - 1] + oneCount;

    if (boxes[i] === '1') {
      oneCount++;
    }
  }

  oneCount = boxes[len - 1] === '1' ? 1 : 0;

  for (let i = len - 2; i >= 0; i--) {
    rightToLeftDp[i] = rightToLeftDp[i + 1] + oneCount;

    if (boxes[i] === '1') {
      oneCount++;
    }
  }

  const ret = [];

  for (let i = 0; i < len; i++) {
    ret.push(leftToRightDp[i] + rightToLeftDp[i]);
  }

  return ret;
};

console.log(minOperations('001011'));
